tests/query_workflow_test.go

```diff
@@ -357,3 +368,121 @@ func (s *QueryWorkflowSuite) TestQueryWorkflow_ClosedWithoutWorkflowTaskStarted(
 	s.Error(err)
 	s.ErrorContains(err, consts.ErrWorkflowClosedBeforeWorkflowTaskStarted.Error())
 }
+
+func (s *QueryWorkflowSuite) TestQueryWorkflow_FailurePropagated() {
+	ctx := testcore.NewContext()
+	taskQueue := testcore.RandomizeStr(s.T().Name())
+
+	workflowRun, err := s.SdkClient().ExecuteWorkflow(ctx, client.StartWorkflowOptions{TaskQueue: taskQueue}, "workflow")
+	s.NoError(err)
+
+	// Create a channel for errors generated in background goroutines.
+	errChan := make(chan error, 1)
+
+	// Query the workflow in the background to have the query delivered with the first workflow task in the Queries map.
+	go func() {
+		_, err = s.FrontendClient().QueryWorkflow(ctx, &workflowservice.QueryWorkflowRequest{
+			Namespace: s.Namespace(),
+			Execution: &common.WorkflowExecution{
+				WorkflowId: workflowRun.GetID(),
+			},
+			Query: &querypb.WorkflowQuery{
+				QueryType: "dont-care",
+			},
+		})
+		errChan <- err
+	}()
+
+	// Hope that 3 seconds will be enough for history to record the query and attach it to the pending workflow task.
+	// There's really no other way to ensure that the query is included in the task unfortunately.
+	util.InterruptibleSleep(ctx, 3*time.Second)
+
+	task, err := s.FrontendClient().PollWorkflowTaskQueue(ctx, &workflowservice.PollWorkflowTaskQueueRequest{
+		Namespace: s.Namespace(),
+		TaskQueue: &taskqueuepb.TaskQueue{Name: taskQueue},
+		Identity:  s.T().Name(),
+	})
+
+	s.Len(task.Queries, 1)
+	qKey := slices.Collect(maps.Keys(task.Queries))[0]
+
+	_, err = s.FrontendClient().RespondWorkflowTaskCompleted(ctx, &workflowservice.RespondWorkflowTaskCompletedRequest{
+		TaskToken: task.TaskToken,
+		Identity:  s.T().Name(),
+		QueryResults: map[string]*querypb.WorkflowQueryResult{
+			qKey: {
+				ResultType:   enumspb.QUERY_RESULT_TYPE_FAILED,
+				ErrorMessage: "my error message",
+				Failure: &failure.Failure{
+					Message: "my failure error message",
+				},
+			},
+		},
+		Commands: []*commandpb.Command{
+			{
+				CommandType: enumspb.COMMAND_TYPE_COMPLETE_WORKFLOW_EXECUTION,
+				Attributes:  &commandpb.Command_CompleteWorkflowExecutionCommandAttributes{},
+			},
+		},
+	})
+	s.NoError(err)
+
+	select {
+	case err = <-errChan:
+	case <-ctx.Done():
+		// Abort and fail the test.
+		s.NoError(ctx.Err())
+	}
+
```

Comment:
Could you add a small comment explaining the purpose of the second half of the test? I don't quite see the difference to the first half.